File: apps/worker/src/batch.ts
import { writeFileSync } from "node:fs";

export function workerOptions(env = process.env) {
  const mode = env.WORKER_MODE ?? "continuous";
  if (!["continuous", "batch"].includes(mode)) throw new Error("invalid_worker_mode");
  function limit(name: string, fallback: number, maximum: number) {
    const value = env[name] ?? String(fallback);
    if (!/^\d+$/.test(value) || Number(value) < 1 || Number(value) > maximum) throw new Error("invalid_worker_limits");
    return Number(value);
  }
  return {
    batch: mode === "batch",
    maxTasks: limit("WORKER_MAX_TASKS", 3, 100),
    maxSeconds: limit("WORKER_MAX_SECONDS", 600, 3600),
  };
}

export class BatchRun {
  readonly started = performance.now();
  readonly options = workerOptions();
  claimed = 0;
  activeTaskId: string | null = null;
  tasks: Array<{ id: string; status: string }> = [];
  reason = "stopped";
  canClaim() {
    if (!this.options.batch) return true;
    if (this.claimed >= this.options.maxTasks) {
      this.reason = "task_limit";
      return false;
    }
    if ((performance.now() - this.started) / 1000 >= this.options.maxSeconds) {
      this.reason = "duration_limit";
      return false;
    }
    return true;
  }
  report(reason = this.reason) {
    const result = {
      worker: "typescript",
      mode: this.options.batch ? "batch" : "continuous",
      reason,
      claimed: this.claimed,
      activeTaskId: this.activeTaskId,
      tasks: this.tasks,
      elapsedSeconds: Math.round((performance.now() - this.started) / 1000),
      recoveryPending: this.activeTaskId !== null,
    };
    if (process.env.WORKER_REPORT_PATH) writeFileSync(process.env.WORKER_REPORT_PATH, JSON.stringify(result));
    console.log(JSON.stringify(result));
  }
  watchdog() {
    if (!this.options.batch) return undefined;
    return setTimeout(() => {
      // Do not release a lease while an uncancellable parser/transaction might still write.
      // Stop this process; a later run recovers the real expired lease through claim_task.
      try {
        this.report("duration_limit");
      } finally {
        process.exit(this.activeTaskId ? 75 : 0);
      }
    }, this.options.maxSeconds * 1000);
  }
}

File: apps/openresults-worker/batch.py
"""Bounded runner accounting; reports never contain payloads, URLs or credentials."""
import json
import os
import threading
import time
from pathlib import Path


def options(env=None):
    env = os.environ if env is None else env
    mode = env.get('WORKER_MODE', 'continuous')
    if mode not in ['batch', 'continuous']:
        raise ValueError('invalid_worker_mode')
    def limit(name, default, maximum):
        value = env.get(name, str(default))
        if not value.isascii() or not value.isdigit() or not 1 <= int(value) <= maximum:
            raise ValueError('invalid_worker_limits')
        return int(value)
    return mode == 'batch', limit('WORKER_MAX_TASKS', 3, 100), limit('WORKER_MAX_SECONDS', 600, 3600)


class BatchRun:
    def __init__(self):
        self.batch, self.max_tasks, self.max_seconds = options()
        self.started = time.monotonic()
        self.claimed = 0
        self.active_task_id = None
        self.tasks = []
        self.reason = 'stopped'

    def can_claim(self):
        if not self.batch:
            return True
        if self.claimed >= self.max_tasks:
            self.reason = 'task_limit'
            return False
        if time.monotonic() - self.started >= self.max_seconds:
            self.reason = 'duration_limit'
            return False
        return True

    def report(self, reason=None):
        result = dict(worker='python', mode='batch' if self.batch else 'continuous',
                      reason=reason or self.reason, claimed=self.claimed,
                      activeTaskId=self.active_task_id, tasks=list(self.tasks),
                      elapsedSeconds=round(time.monotonic() - self.started),
                      recoveryPending=self.active_task_id is not None)
        path = os.environ.get('WORKER_REPORT_PATH')
        if path:
            Path(path).write_text(json.dumps(result), encoding='utf-8')
        print(json.dumps(result), flush=True)

    def watchdog(self):
        if not self.batch:
            return None
        def expire():
            # asyncio cancellation cannot stop a live psycopg/publisher thread.
            # Exit the process without releasing the lease; recovery stays fenced.
            try: self.report('duration_limit')
            finally: os._exit(75 if self.active_task_id else 0)
        timer = threading.Timer(self.max_seconds, expire)
        timer.daemon = True
        timer.start()
        return timer
